feat: private range proof

thomas-waite opened the pull request:
## Summary
This PR implements a zero-knowledge proof that a confidential digital asset has a value greater or less than another confidential digital asset. This is a private range range proof.

This PR includes: JavaScript proof construction and verification code, ABI input encoding, smart contract zk proof validator, smart contract output ABI encoding and tests. 

## What I would like feedback on/where to focus
1) `(packages/aztec.js/src/proof/privateRange)': The API for `privateRange.encodePrivateRangeTransaction()` - it's inputs are formatted slightly differently to the other proofs. Specifically, rather than taking `inputNotes` and `outputNotes` as inputs, in the context of this proof I thought it would be more intuitive for the user to pass in `originalNote`, `comparisonNote` and `utilityNote`

2) `(packages/aztec.js/src/proof/privateRange)': I have not mixed `publicOwner`, `publicValue` into the hash used to construct the challenge (`proofUtils.constructChallenge()`) as they are constant for all possible instantations of this proof. Thoughts on this and other variables to mix in? (currently put senderAddress, notes and blindingFactors in)

3) `(packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol)`: Due to the balancing relation equation in this proof, constructing a less than proof (rather than a greater than proof) I believe would require the user to create a greater than proof to the same effect
- e.g. if you want to prove that k_1 < k_2, then you would have to use this proof to show that k_2 > k_1. With my current proof relation, I don't believe I can directly show that k_1 < k_2, like you would be able to in a public range proof

4) `(packages/protocol/contracts/ACE/validators/privateRange/PrivateRangeABIEncoder.sol)`: Currently, in the `bytes proofOutput` object I am returning a non-zero `outputNotes` field - I am returning the utility note. The alternative is to return an empty field, as this output/utility note will not be used for anything after proving the greater than/less than property. Interested in thoughts on this? Returning an output note is consistent with the other proofs and function setups 



## Packages updated
- [x] aztec.js - Javascript proof construction and verification code, ABI encoding public range proof input coder and tests
- [x] protocol - zero-knowledge proof validation smart contract, ABI encoding contract and tests

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

_aztec.js_
- [x] Javascript proof construction code
- [x] Tests for well formed proof data
- [x] Javascript proof verification code
- [x] Tests for successful validation and checking of failure states
- [x] ABI input encoding 
- [x] Tests for ABI input encoding

_protocol_
- [x] zk proof validator contract
- [x] Basic tests for zk proof validator
- [x] ABI output encoding contract
- [x] Basic tests for ABI output encoding contract
- [x] Extensive tests for zk validator

thomas-waite committed:
feat(aztec.js): implement private range proof construction

thomas-waite committed:
feat(aztec.js): implement private range JS proof verification code

thomas-waite committed:
refactor(aztec.js): export private range proof construction module

thomas-waite committed:
test(aztec.js): tests for private range proof construction

thomas-waite committed:
test(aztec.js): tests for private range proof verification

thomas-waite committed:
feat(aztec.js): method for ABI input encoding private range

thomas-waite committed:
feat(aztec.js): add privateRange option in encode() method

thomas-waite committed:
test(aztec.js): tests for privateRange proof encoderFactory methods

thomas-waite committed:
feat(protocol): implement smart contract interface for PrivateRange.sol

thomas-waite committed:
feat(protocol): first draft of private range zk validator contract

thomas-waite committed:
feat(protocol): success test for PrivateRange.sol

thomas-waite committed:
refactor(protocol): make trace.sh executable

thomas-waite committed:
style: lint and prettier fixes throughout

thomas-waite committed:
test(protocol): add failure tests for private range proof

thomas-waite committed:
refactor(protocol): implement correct relation between k_bar factors

thomas-waite committed:
refactor: change interface return, remove DOMAIN_HASH,remove console.log

PrivateRangeInterface.sol - remove the EIP712 domain hash, and change return variable type from bytes32 to bool
aztec.js/../proof/privateRange/index.js - remove unnecessary console.log()
Prettier fixes

thomas-waite committed:
feat(protocol): implement PrivateRangeABIEncoder.sol

thomas-waite committed:
refactor(protocol): refactor PrivateRange.sol to make use of ABI encoder

thomas-waite committed:
test(protocol): implement test for PrivateRangeABIEncoder.sol

thomas-waite committed:
refactor(aztec.js): call the 'privateRange' encoding rather than 'burn'

thomas-waite committed:
style(protocol): apply prettier changes

thomas-waite committed:
test(protocol): implement further success, failure private range tests

thomas-waite committed:
test(aztec.js): remove redundant on curve checking code

proofUtils.parseInputs(notes, sender) performs the same functionality

thomas-waite committed:
refactor(aztec.js): avoid sending redundant info, set k_3 to 0

thomas-waite committed:
test(protocol): failure tests for variables not in challenge

thomas-waite committed:
test(aztec.js): make tests work with new note construction API

thomas-waite committed:
test(protocol): update tests to use new note construction API

thomas-waite committed:
refactor: update privateRange.constructProof() method for consistency

Done to be consistent with other proofs. Now takes spreaded inputNotes and outputNotes, rather than original Note, comparison note and utilityNote

thomas-waite committed:
refactor(aztec.js): generate proof data for > max num notes

Done to enable a failure test for num notes > max num notes

thomas-waite committed:
test(protocol): add failure test for too many notes

thomas-waite committed:
refactor(protocol): set n=3, rather than load from the calldata

thomas-waite committed:
style: doc updates, comment fixes/removals

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+const Keccak = require('../../keccak');
+
+const privateRange = {};
+privateRange.verifier = verifier;
```
We should avoid assigning members of one module into members of another module, as it opens up the possibility of a code injection attack (if one can tamper with the client, you can swap this method out with a second method that, for example, has the same functionality but records every input parameter when this method is called)

ArnSch replied:
It's not clear to me that this is a problem introduced by this pattern. Even if the fn definition for `verifier` happened in the privateRange module, someone with the ability to tamper with the client could simply define a file which requires `privateRange` and redefines `privateRange.verifier`, and as long as that file is loaded prior to the actual use of `privateRange.verifier` in the client, the malicious, replaced fn will be called.

In fact, we use this pattern in tests to replace the method which gets a point from the trusted setup database.

If I misunderstood the attack you were thinking about, could you clarify? Otherwise I think duplicating the `verifier` code is not worth it.

thomas-waite replied:
As discussed, leaving as is

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+
+    proofUtils.parseInputs(notes, sender);
+
+    notes.forEach((note) => {
```
Do we check that there are only 3 notes in this method?

thomas-waite replied:
Good point, added in a check

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+                .redAdd(blindingFactor.bk)
+                .fromRed();
+        } else {
+            kBar = padLeft(new BN(crypto.randomBytes(32), 16).umod(bn128.curve.n).toString(16), 64);
```
Should this be 0? I don't follow why we set this to be a random variable

thomas-waite replied:
We currently throw an error during proof construction if the scalars are zero. So, setting kBar to 0 here would cause proof construction to fail. 

I agree that this behaviour is desired though, will look into it. Opened an issue to track this

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+            .fromRed();
+
+        return [
+            `0x${padLeft(kBar.toString(16), 64)}`,
```
(this isn't really part of the review, but we should add a TODO task to perform this formatting in the abi encoder)

thomas-waite replied:
Agreed, added it to my TODO list

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+    });
+    return {
+        proofData,
+        challenge: `0x${padLeft(challenge.toString(16), 64)}`,
```
Do we restrict challenge to be modulo the generator order? We should be validating this is the case in the smart contract

thomas-waite replied:
When I load the the challenge variable from calldata in the validator contract, `PrivateRange.sol`, I load it under modular arithmetic by doing the following:

`let challenge := mod(calldataload(0x124), gen_order)`

calldata location 0x124 corresponds to the challenge. 

I think this covers it, but if not let me know

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+ * @returns {Object} AZTEC proof data and expected output
+ */
+privateRange.encodePrivateRangeTransaction = ({ originalNote, comparisonNote, utilityNote, senderAddress }) => {
+    const inputNotes = [...originalNote, ...comparisonNote];
```
We should probably make the utility note inside `privateRange.constructProof`. This utility note should be created explicitly for this proof and then not re-used. If we allow the note to be passed as an input parameter it opens up the possibility that an implementor re-uses notes (and note secrets) and we want to avoid that possibility.  
  
It also makes the interface a bit simpler as the user has to keep track of fewer notes.

thomas-waite replied:
I agree about constructing the `utilityNote` internally to the proof construction. 

I have implemented, but have constructed inside `privateRange.encodePrivateRangeTransaction()` instead - I found it more straightforward to then perform the required tests. It also means that the `privateRange.constructProof()` method is solely focused on implementing the cryptography, rather than also constructing an additional note.

Constructed a `helper` function, with a `helpers.constructUtilityNote()` method to do this.

Interested in your thoughts

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+
+    const blindingFactors = privateRange.constructBlindingFactors(notes, rollingHash);
+
+    const challenge = proofUtils.computeChallenge(sender, notes, blindingFactors);
```
I think we should include `kPublic` and `publicOwner` in this challenge, even though they are 0. We want to have a consistent 'interface' for constructing AZTEC challenges, this will become more relevant when we start to implement virtual machine functionality

thomas-waite replied:
Done

zac-williamson commented on packages/aztec.js/src/proof/privateRange/verifier.js:
```diff
@@ -0,0 +1,125 @@
+ * @returns {boolean} valid - boolean that describes whether the proof verification is valid
+ * @returns {option} errors - an array of all errors that were caught
+ */
+verifier.verifyProof = (proofData, challengeHex, sender) => {
```
:thumbsup:

zac-williamson commented on packages/aztec.js/src/proof/privateRange/index.js:
```diff
@@ -0,0 +1,161 @@
+/**
+ * Constructs AZTEC private range zero-knowledge proofs
+ *
+ * @module privateRange
```
We should probably add a bit of documentation here about how this works, specifically with regards to the input note and the comparison note. These notes need to be switched around, to convert this proof from a 'greater than or equals' proof to a 'less than or equals' proof, we should mention this

thomas-waite replied:
Done

zac-williamson commented on packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol:
```diff
@@ -0,0 +1,355 @@
+    
+        // if we've reached here, we've validated the private range proof and haven't thrown an error.
+        // Encode the output according to the ACE standard and exit.
+        PrivateRangeABIEncoder.encodeAndExit();
```
we should probably hoist this to the start of the fallback function. i.e. just after validatePrivateRange() is called. I know this isn't something we do in our other contracts, we should probably change all of them to make the program flow a bit clearer

thomas-waite replied:
Good idea, opening an issue to deal with this across all our validator contracts

zac-williamson commented on packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol:
```diff
@@ -0,0 +1,355 @@
+                    validateCommitment(noteIndex, k, a)
+
+                    // Set k = kx_j, a = ax_j, c = cx_j, where j = i - (m+1)
+                    if gt(i, 0x00) {
```
(small optimiation, but we can replace `if gt(i, 0x00)` with `if i`. The `jumpi(x, location)` opcode will execute the jump instruction if `x` is anything but zero.

thomas-waite replied:
Nice, done

zac-williamson commented on packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol:
```diff
@@ -0,0 +1,355 @@
+                    // If i > m + 1 (i.e. subsequent output notes)
+                    // then we add \sigma^{-c} and \sigma_{acc} and store result at \sigma_{acc} (0x1e0:0x200)
+                    // we then calculate \gamma^{cx} and add into \gamma_{acc}
+                    if gt(i, m) {
```
m = 2 and n = 3, so this never gets called.

We should discuss in person but I think we should include both all 3 notes in the pairing check. Depending on whether this is a 'greater than' or 'less than' proof, the input and comparison notes get swapped around.

I think we should always include the comparison note in the range proof check, as the contract calling this range proof check may not have any agency over the structure of this note (it could be prover defined). So we should include all 3 notes in the pairing check.

thomas-waite replied:
As discussed + reviewed, done

zac-williamson reviewed (changes requested):
Heya! Great job on this, I have a couple of comments surrounding the structure of the zero knowledge proof, but overall this is a solid implementation of our range proof.

thomas-waite committed:
feat: add proofID constant for private range proof

thomas-waite committed:
test(protocol): privateRange test for ace.validateProof()

thomas-waite committed:
style(protocol): apply prettier changes

thomas-waite committed:
refactor(aztec.js): implement PR feedback

add a check into proof construction to check number of input notes

thomas-waite committed:
refactor: implement PR feedback

add Kpublic and publicOwner into the hash used to construct the challenge. Done for standardisation
update tests accordingly, and add in 2 new tests to validate failure when kPublic or publicOwner not mixed into hash

thomas-waite committed:
refactor(protocol): implement PR feedback

small optimisation, replace gt() with an if

thomas-waite committed:
refactor(aztec.js): construct utility note internally

done to make interface simpler and more secure
wrote a helper function constructUtilityNote()

thomas-waite committed:
test(aztec.js): update tests with internal utility note construction

thomas-waite committed:
test(protocol): implement PR feedback

update protocol tests to use new interface where utility note is created internally

thomas-waite committed:
refactor(protocol): implement PR feedback

include all notes in the bilinear pairing comparison

thomas-waite committed:
test: update tests to work with privateRange interface

thomas-waite committed:
fix(aztec.js): remove < else if in helper.constructUtilityNote()

there should not be a scenario where originalValue < comparisonValue - as the user should be
swapping around the note values to construct a less than relation

thomas-waite committed:
refactor: add errorType in dev-utils, eslint fixes

thomas-waite committed:
refactor: add in option of manual utilityNote construction

edit the interface of privateRange.encodePrivateRangeTransaction()

thomas-waite committed:
style(aztec.js): update docs to explain default+manual utility note flow
